### ISSUE TITLE
Steam Deck support

### ISSUE THREAD

**Issue by Atemu** [NixOS/nix#7173]
**Is your feature request related to a problem? Please describe.**

It is currently not possible to install Nix on the Steam Deck using the regular installer due to its special immutable rootfs setup.

Valve has left us with quite some options however:

* We have superuser privileges
* /etc is permanently writeable, including fstab and systemd units
* Home is a bog standard read-write Unix filesystem (ext4) with quite a bit of space (depending on the model)

**Describe the solution you'd like**

Ideally we could have Valve add an empty `/nix/` directory to their SteamOS image. This way we could bind-mount a directory inside home onto `/nix/` via a simple systemd mount unit and Nix could operate as usual.

IIRC we have tried to reach out to Apple regarding `/nix` on macOS before but that wasn't very fruitful (heh).  
Going by their past behaviour, I imagine Valve might be much more cooperative on such matters.

Perhaps we could even reach out to them in an "official" manner via the NixOS foundation.  
The Guix folks might also have an interest in joining our request with one for a `/gnu/` directory.

**Describe alternatives you've considered**

It is probably possible to achieve a rw `/nix/` without the help of Valve using overlayfs, selectively mounting `/usr/` etc. onto a custom tmpfs root or worse hacks but those might not even be possible or might be fragile.

**Additional context**

The Steam Deck has limited possibilities in extending it with your own software due to its read-only rootfs. You can't just install Arch packages via pacman for instance.  
Flatpak is installed and enabled by default and it covers GUI apps fairly well but it doesn't cover CLI apps at all.

Nix could really shine here.

**Comment by Enzime** [comment]
I believe you can just run the following commands:

```
$ sudo steamos-readonly disable
$ sudo mkdir /nix
$ sudo steamos-readonly enable
```

However I believe this would get wiped out by Steam Deck updates, but it would probably be possible to write a systemd unit to set this up every boot

**Comment by Atemu** [comment]
Namespaced "solutions" are not good. They come with a whole range of issues. Most immediate of which is that you always need to enter a namespace before being able to run anything installed via Nix. This causes all kinds of headaches and that is not acceptable.

**Comment by L-as** [comment]
:+1: for the namespacing solution, I don't think there are much better solutions for this without heavily reworking how Nix works.

**Comment by e-minguez** [comment]
https://determinate.systems/posts/nix-on-the-steam-deck

This (bind mounts and systemd units) is much more elegant than Valve adding an empty /nix dir for no reason.

**Comment by Atemu** [comment]
I was looking into their solution just yesterday, great to have a write-up.

I don't think disabling the read-only root is in any way elegant but, done in a service like this, it is a quick and dirty solution that works.

They linked @Xe's article on how to install Tailscale on the Deck though where they describe using systemd-sysext to install the tailscale binary and apparently that works without "cracking the readonly seal". Not sure whether sysext could create `/nix` for us though as it might be limited to stupid FHS dirs such as `/usr` but, should that work, sysext would certainly be an elegant solution; without Valve even needing to lift a finger.

***

Man, it's so great to have a (mostly) bog-standard Linux with bog-standard systemd and all its features on a friggin console.

**Comment by e-minguez** [comment]
> I was looking into their solution just yesterday, great to have a write-up.
> 
> I don't think disabling the read-only root is in any way elegant but, done in a service like this, it is a quick and dirty solution that works.
> 
> They linked @Xe's article on how to install Tailscale on the Deck though where they describe using systemd-sysext to install the tailscale binary and apparently that works without "cracking the readonly seal". Not sure whether sysext could create `/nix` for us though as it might be limited to stupid FHS dirs such as `/usr` but, should that work, sysext would certainly be an elegant solution; without Valve even needing to lift a finger.
> 
> ***
> 
> Man, it's so great to have a (mostly) bog-standard Linux with bog-standard systemd and all its features on a friggin console.

If they add FDE it would be awesome !

**Comment by Hoverbear** [comment]
Unfortunately `systemd-sysext` can't create a `/nix`, I did experiment with this! Unfortunately making the root temporarily writable is the best option I've found. Certainly interested in finding a new idea. :)

**Comment by Nephyrin** [comment]
SteamOS 3.5 will have a `/nix/` directory available.  Currently the nix package is not pre-installed, however.

**Comment by Scrumplex** [comment]
> SteamOS 3.5 will have a `/nix/` directory available. Currently the nix package is not pre-installed, however.

I assume this also means that it will be writable, then? Or will we need to mount our own partition there?

**Comment by Atemu** [comment]
> SteamOS 3.5 will have a `/nix/` directory available. Currently the nix package is not pre-installed, however.

That'd be amazing! Do you have a source for that?

Do they also ship /gnu/?

> I assume this also means that it will be writable, then? Or will we need to mount our own partition there?

I doubt it'd be writeable but this frees us of having to create that directory in the read-only root which is the fragile part. We can bind-mount a rw directory on top at runtime without modifying the data on-disk.

**Comment by Nephyrin** [comment]
> That'd be amazing! Do you have a source for that?

Source is myself, being a SteamOS developer!  It is configured via the `holo-nix-offload` package, which you can peep the source of on the mirror at https://steamdeck-packages.steamos.cloud/archlinux-mirror/sources/jupiter-main/holo-nix-offload-0.1-1.src.tar.gz

If you try the (often unstable!) Main branch of SteamOS you can test it out.

> Do they also ship /gnu/?

Not currently, but would be trivial to add

> I assume this also means that it will be writable, then? Or will we need to mount our own partition there?

It is currently bind-mounted to the data partition at `/home/.steamos/offload/nix/` along with other 'offload' directories that are independent of OS update (unlike the small/volaile `/var` which undergoes a selective migration on update).

I would also be interested in opinions on having the nix arch package installed by default in SteamOS builds.  SteamOS tends to ship a minimal set of OS packages and rebase on arch with month(s) of latency, so it could be detrimental 'helpfully' pre-ship a potentially-old Nix binary.  Does Nix have a good self-replacing-with-nixpkgs-nix story in that case?

**Comment by Hoverbear** [comment]
@Nephyrin You might be able to ship a minimal nix which users could `nix upgrade-nix` to get a recent copy?

A side note, we added support for the new `/home/.steamos/offload/nix` recently in our `nix-installer` crate (https://github.com/DeterminateSystems/nix-installer/pull/495) and it works really great! We did have an issue someone raised about wanting to use `nix` from an sdcard though (https://github.com/DeterminateSystems/nix-installer/issues/509), so we may need to tweak support. Feel free to try it out (off the `main` branch at this moment) `curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix/branch/main | sh -s -- install`

**Comment by Atemu** [comment]
> Source is myself, being a SteamOS developer!

Oh! Well, that's a rather good source then :D

Thanks a bunch for considering us <3

> If you try the (often unstable!) Main branch of SteamOS you can test it out.

I'm personally short in time for the next ~month but I'm sure others will try it on their Decks.

> Not currently, but would be trivial to add

I'm sure the GUIX folks would appreciate that!

> It is currently bind-mounted to the data partition at `/home/.steamos/offload/nix/` along with other 'offload' directories that are independent of OS update

That sounds like a sane strategy. It'd be great if the user could choose where the offload directory is located physically like @Hoverbear mentioned; some users might want Nix on the SD card.  
Though technical users could of course just set up fstab/mount units to accommodate that and Nix users are very likely to be technical.

> I would also be interested in opinions on having the nix arch package installed by default in SteamOS builds.

This could be done but I don't think it's necessary. Unless you plan to make use of Nix within SteamOS, I don't think every SteamOS user needs Nix. The Nix package itself has a significant size and pulls in quite a few dependencies, some of which even most Nix users do not need (such as the AWS SDK).  
The ones who do need/want Nix will be able to install it themselves just fine as long as `/nix/` exists and `/etc` can be modified for environment glue.

Nix shouldn't really be managed by another package manager anyways IMHO; it's its own package manager afterall and stays out of the way of the "host" system as much as it can.

>  SteamOS tends to ship a minimal set of OS packages and rebase on arch with month(s) of latency, so it could be detrimental 'helpfully' pre-ship a potentially-old Nix binary.

I don't think that'd cause any real issues. Nix is quite conservative when it comes to compatibility and stays backwards compatible. You can still eval the current Nixpkgs with a Nix build from 2019 and I'm sure even earlier builds work too.

Just a few months wouldn't be anything to be concerned about, other than security etc. which always applies.

**Comment by Ericson2314** [comment]
Oh good. I've been hoping Nix would show up on Valve's radar for a very long time. 

> Does Nix have a good self-replacing-with-nixpkgs-nix story in that case?

Potentially :). There is an experimental `nix upgrade-nix` command, for example.

You can ship a single statically linked binary of Nix and it works, so in some sense this is thankfully not too hard a problem. But with configuring the system to be able to use all the features it gets harder. Thankfully there are new/upcoming features like https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-auto-allocate-uids that obviate needing to configure the underlying system; I certainly hope we can continue down that route so installing Nix becomes successively more trivial.

@Nephyrin I am happy to answer any further questions / otherwise assist in any way I can. I am sure others from the Nix team besides myself would similarly be interested too.

**Comment by Enzime** [comment]
As a Steam Deck owner I would love to see Nix pre installed, I think there will definitely be utility as users can run programs without having to install them. The good thing about SteamOS being based on Arch is they can use the Arch Nix package as an easy bootstrap.

I just tested Nix on Arch, and I agree one potential blocker would be that Nix pulls in AWS SDK as a dependency which is 500MB, the rest of Nix and its dependencies are only like 100MB. I think it would be worth making AWS SDK an optional dependency as from what I saw it was only used for S3 binary stores, and it would reduce the upfront storage cost for users.

Another issue is that `nix upgrade-nix` only works when Nix is installed from a profile, whereas the Arch package installs it to `/usr/bin` and the Arch Nix Daemon service uses `/usr/bin/nix`. I’m not sure what the best solution for improving the `upgrade-nix` on Arch (and SteamOS) is.

If there’s interest on the Nix team side, I’d be happy to work on these issues.

**Comment by Ericson2314** [comment]
The AWS SDK is already an optional dependency, actually. See the various `ENABLE_S3` in the Nix codebase.

I don't think it makes sense for Nix to know how to *un*install the Arch Nix package, but any version of Nix is capable of installing Nix via Nix, and we could try to somehow make sure the Arch is "shadowed" the newly installed Nix (e.g. `PATH` order and whatnot).

Still, such shadowing is a bit fragile. Do you have a link to the Arch Nix package? I'd be curious to see what it does and doesn't do. That will help inform how much value we get out of it versus something else.

**Comment by Enzime** [comment]
Oh, I totally missed that, I wonder if people would be willing to change the default to false to reduce the closure size.


Here's the source files for the Nix package:

https://gitlab.archlinux.org/archlinux/packaging/packages/nix

**Comment by Ericson2314** [comment]
Thanks @Enzime.

I'll go through the files in here

- https://gitlab.archlinux.org/archlinux/packaging/packages/nix/-/blob/main/nix.conf

Needed for `build-users-group` today. Could do `auto-allocate-uids` instead, but still need file for that.

Easy to create `/etc/nix/nix.conf` ourselves, however.

- https://gitlab.archlinux.org/archlinux/packaging/packages/nix/-/blob/main/sysusers.conf

Setting up build users and groups. Can drop entirely with `auto-allocate-uids`.

- https://gitlab.archlinux.org/archlinux/packaging/packages/nix/-/blob/main/tmpfiles.conf

Setting up some state directories. Still need those directories, but I think can be done by Nix itself on startup. File shouldn't be needed.

- https://gitlab.archlinux.org/archlinux/packaging/packages/nix/-/blob/main/user.environment.conf

Unneeded. The users of Nix can set `PATH` themselves. `NIX_REMOTE=daemon` will be auto-detected. `NIX_PATH` is purely optional, can also be set by users if they want it.

- https://gitlab.archlinux.org/archlinux/packaging/packages/nix/-/blob/main/user.tmpfiles.conf

Unneeded. Since Nix 2.15 all of this stuff is within the user's home directory, and will be created as needed on demand.

- https://gitlab.archlinux.org/archlinux/packaging/packages/nix/-/blob/main/PKGBUILD

Main file. Nothing that interesting but installing the other files.

**Comment by Ericson2314** [comment]
So in conclusion, I don't think there is much extra value being gotten from the Arch Nix installer. The main point of system integration we don't currently have a replacement for is integrating with the systemd configuration, but with https://github.com/NixOS/nix/pull/5380 even the Nix daemon can just be run as non-root, making the systemd configuration a good bit lower stakes.

**Comment by Atemu** [comment]
@Ericson2314 just so there's no misunderstanding, we can freely modify `/etc/` on SteamOS. All the stuff the regular Nix installer configures (users, shells, systemd units, nix.conf etc.) just works like it does everywhere else. The only thing we're unable to do (without disabling readonly) is the creation of `/nix/` which apparently SteamOS now does for us.

**Comment by Ericson2314** [comment]
Yes chain of trusts with installers can definitely work.

Also to @Atemu's point, it is precisely because the Arch installer could do arbitrary things that commands like `nix upgrade-nix` are wary of clobbering it. A pre-install that puts nix in `/nix/store/...` not `/usr/bin` will not trigger those cautious aborts, and self-upgrade path works just fine in that case.

In a way, these are not that different. A signed/trampolined installer or pre-installed Nix without involving Arch both lead to basically the same place, and if a user wants to do a self-upgrade to a newer version of Nix form there, that is the same process too.

**Comment by bertogg** [comment]
@Atemu 

> we can freely modify `/etc/` on SteamOS. All the stuff the regular Nix installer configures (users, shells, systemd units, nix.conf etc.) just works like it does everywhere else.

I think that there's an additional problem: the installer script does put the systemd unit files in `/etc/systemd`, but they are symlinks to the actual files under `/nix`. Those files are however not available until systemd mounts `/nix`, so the symlinks are dead and therefore ignored when systemd reads its configuration at boot time (it does work if you reload the systemd configuration after `/nix` is mounted).

This does not affect single-user mode as far as I can tell, and Nix seems to work just fine in that scenario.

However I suppose that it would also be nice if both the `root` and `deck` users can run software installed with Nix. Is that only possible if Nix is running in multi-user mode, or is there any other alternative?

**Comment by Atemu** [comment]
Thank you for pointing that out @bertogg. That would indeed not work.

Although I see little reason for root to run nix tools, multi-user Nix is the only well supported option, so we do need multi-user.

NixOS gets around this issue by mounting `/nix` in the initrd IIRC but I'm not familiar with how that's done in the new systemd initrd.

A simple way around this issue would be to copy the file rather than symlinking. A similar thing is done on macOS for the `org.nixos.nix-daemon.plist` because `/nix` is mounted separately too. I'm not sure how `nix upgrade-nix` works in this case though as it must have a hook to update the file?

**Comment by bertogg** [comment]
The general issue would be "Support installations where `/nix` is not in the root filesystem", as this could conceivably happen anywhere.

At first sight I can also think of two possible solutions:

1. Copy the files as you say and make sure that they are updated if Nix is upgraded.
2. Reload the systemd configuration (the solution proposed in [this post](https://determinate.systems/posts/nix-on-the-steam-deck) does that), but that might be tricky to implement in a generic way (i.e. not only for SteamOS). I also don't like the idea of a systemd unit running `systemctl daemon-reload`, maybe it's supported but intuitively it feels hackish and error prone.

**Comment by Nephyrin** [comment]
> Do they also ship /gnu/?

A quick check here shows that the guix installer needs a bit more interaction with the system than just a `/gnu` mount available, so it's not clear if this would be useful standalone.  If any guix folks are interested in pushing on this let us know

**Comment by alkazar** [comment]
ChimeraOS dev here. This is a really interesting idea. Would it be helpful to mount `/nix` as a read-write BTRFS subvolume? We do this for `/home` and `/var` already.

ChimeraOS is also based on Arch, but pushes updates more frequently. Usually once a month. In such a case, would it be useful to include the `nix` Arch package? Or would it still get in the way?

**Comment by SuperSandro2000** [comment]
> [gitlab.archlinux.org/archlinux/packaging/packages/nix/-/blob/main/nix.conf](https://gitlab.archlinux.org/archlinux/packaging/packages/nix/-/blob/main/nix.conf)

Since contributing to arch is not as easy as just opening a MR on their GitLab can someone remove the line suggesting that disabling sandbox is a good idea? Thanks!

**Comment by bertogg** [comment]
@nckx if you don't mind please share the link to guix-devel.

Problems that I noticed when installing Guix on the Steam Deck (assuming that `/gnu` is a bind mount to a writable place with enough empty space):

1. Unlike Nix, Guix expects that the root filesystem is writable and installs files (symlinks) such as `/usr/local/bin/guix`, `/usr/local/share/info/*` and a couple of shell completion scripts (`/usr/share/fish/vendor_completions.d/guix.fish`, `/usr/share/zsh/site-functions/_guix`).  I haven't seen a way to opt out of those files.

In addition to that, the tarball with the distribution (`guix-binary-*.tar.xz`) is also a problem because it is unpacked in the root directory and contains a `./` entry:

```
$ tar --extract --file guix-binary-1.4.0.x86_64-linux.tar.xz -C /
tar: .: Cannot utime: Read-only file system
tar: .: Cannot change ownership to uid 0, gid 0: Read-only file system
tar: Exiting with failure status due to previous errors
```

This can be worked around by making the rootfs read-write before running `guix-install.sh`. However: 

2. A SteamOS system update replaces the contents of the root filesystem (`/etc` is preserved) so all those extra files installed by Guix disappear. The `/usr/local/bin/guix` symlink is the main problem because it will be gone after an update.

On the positive side the systemd units installed by Guix are not symlinks so they don't have the problem mentioned earlier  in this thread: https://github.com/NixOS/nix/issues/7173#issuecomment-1609822014.

**Comment by podiki** [comment]
Hello! Also from Guix and I do have a Steam Deck. I even booted a Guix USB stick when I first got it but haven't tried putting Guix as a package manager on there. Would be happy to help test and contribute on this front.

**Comment by nckx** [comment]
@bertogg It was an [e-mail address](https://lists.gnu.org/mailman/listinfo/guix-devel) all along!  We're old-school.

> Unlike Nix, Guix expects that the root filesystem is writable and installs files

These aren't *strictly* necessary: they are there to allow each user to bootstrap their first ‘guix pull’.

Does SteamOS have any root-writable directory in `$PATH`?

> the tarball […] `contains` a ./ entry

I think/hope the [fix for this was trivial](https://git.savannah.gnu.org/cgit/guix.git/commit/?id=ff23e34f564820172964e9b9de096432b8f7d006).

**Comment by samueldr** [comment]
Hi, uh...

You should all know that I have had an official word about this for a *long while*.

Vagueposted about the offload package appearance here

> 20230426.1000 might be a new milestone for Nix adjacent things... Maybe more about that later...  
> **[7:44 PM · Apr 26, 2023](https://twitter.com/samueldr/status/1651371723588091908)**

 - https://twitter.com/samueldr/status/1651371723588091908

Then reached-out for a word, hopefully I'm not overreaching by sharing these quotes outright. Emphasis mine.

*Tue, May 2, 2023*

> To give some more background here, we have been exploring/discussing what a system with the robustness guarantees of a read-only A/B system but the flexibility of a more traditional linux environment might look like.  For instance, **tools like Nix** and Distrobox **offer approaches to separating the base OS from developer/tooling/etc. environments**, while Flatpak and AppImage (and I guess Snap) offer a path for distributing more end-user oriented software outside the system package manager, and so on.  The hope is that **making tweaks to improve compatibility with such tools encourages the community to explore their usage** more and unblocks some use-cases in the short term.
>
> (It's been noted that Nix is actually a good vector for installing podman/distrobox in a rootless manner, as well)
>
> In that light, the `holo-nix-offload` package **was intended to be a minor change that allows the /nix toplevel directory to be remapped to the home/data partition, removing the need to toggle the read-only state of the rootfs** or use nonstandard nix roots.  It was not intended to be everything required for Nix usage, **nor an official promise of any future support**, just a small tweak to make it easier for those who do wish to play with Nix on their devices without OS updates interfering.  You're correct that currently we don't currently pre-install the Nix package, so **users would still need to use the out-of-band installer** or otherwise.
>
> If we were to communicate this publicly, it would likely just be a patch note under a 'for developers' section roughly noting the above, and making clear that **it is not equivalent to official/full/proper Nix support**.

So it's not *support* by Valve, but what I would call an *affordance*. They make it so you can have your cake and eat it too. In other words, keep using their read-only A/B system, and yet still make it possible to use Nix ecosystem things.

A bit off-topic for the Nix repository, but for anything in the Nix/NixOS ecosystem related to Steam Deck and Steam OS, you should ping at least me. There is also [Jovian-NixOS](https://github.com/Jovian-Experiments/Jovian-NixOS/) that may be relevant in some form here, but a different thing altogether. Apparently [I've not been loud enough about this project](https://ap.samueldr.com/notice/AZGjnC3llfRq2raXGi)*!*

**Comment by bertogg** [comment]
@Atemu

> A simple way around this issue would be to copy the file rather than symlinking.

FWIW the DeterminateSystems Nix installer uses the `daemon-reload` approach:

https://github.com/DeterminateSystems/nix-installer/blob/v0.12.0/src/planner/steam_deck.rs#L270

**Comment by Hoverbear** [comment]
We do! Since the Nix store gets bind mounted, some systemd units aren't available at systemd daemon startup time (the symlinks exist but are dead), but once the mount units fire those symlinks now point to real files, so we need the daemon to reload to find those files and include them!

It feels gross and dirty, but it does work!

**Comment by al12gamer** [comment]
> @Atemu
> 
> > A simple way around this issue would be to copy the file rather than symlinking.
> 
> FWIW the DeterminateSystems Nix installer uses the `daemon-reload` approach:
> 
> https://github.com/DeterminateSystems/nix-installer/blob/v0.12.0/src/planner/steam_deck.rs#L270

So, wait, does this mean I can install nix pkgs fine on steam deck by just running the DeterminateSystems installer?

**Comment by grahamc** [comment]
Yep!

**Comment by al12gamer** [comment]
SteamOS 3.5.5 hit yesterday I believe. I'll have to see soon if this means we can just run the nix env installer.

**Comment by DraconicNEO** [comment]
> SteamOS 3.5.5 hit yesterday I believe. I'll have to see soon if this means we can just run the nix env installer.

Have you tried yet? I know this is relatively new and I haven't seen many developments on it but hoping we'll be able to use Nix in an official capacity that isn't broken by updates soon.

**Comment by bertogg** [comment]
The official Nix installer won't work in daemon mode because of the problem discussed here: https://github.com/NixOS/nix/issues/7173#issuecomment-1609822014

The DeterminateSystems installer does work, and so does the official installer in single-user mode.

**Comment by necrophcodr** [comment]
> The official Nix installer won't work in daemon mode because of the problem discussed here: [#7173 (comment)](https://github.com/NixOS/nix/issues/7173#issuecomment-1609822014)
> 
> The DeterminateSystems installer does work, and so does the official installer in single-user mode.

I've tried the single-user mode of the official installer, and I cannot confirm that it works. What parameters did you specify to make it work at all?

**Comment by Atemu** [comment]
SteamOS has `sudo`. You need to set a password first IIRC. I'm sure there are tonnes of guides on the internet on that.

**Comment by al12gamer** [comment]
> > I've tried the single-user mode of the official installer, and I cannot confirm that it works. What parameters did you specify to make it work at all?
> 
> ```
> (deck@steamdeck ~)$ grep _ID /etc/os-release 
> VARIANT_ID=steamdeck
> VERSION_ID=3.5.5
> BUILD_ID=20231116.2
> 
> (deck@steamdeck ~)$ sudo chown deck:deck /nix
> [sudo] password for deck: 
> 
> (deck@steamdeck ~)$ wget https://nixos.org/nix/install
> [...]
> 
> (deck@steamdeck ~)$ sh ./install --no-daemon
> downloading Nix 2.19.1 binary tarball [...]
> performing a single-user installation of Nix...
> copying Nix to /nix/store.................................................
> installing 'nix-2.19.1'
> building '/nix/store/7kk5lwf8qmnplp54xa6hxzr1kq5zx8jw-user-environment.drv'...
> unpacking channels...
> modifying /home/deck/.bash_profile...
> 
> Installation finished!  To ensure that the necessary environment
> variables are set, either log in again, or type
> 
>   . /home/deck/.nix-profile/etc/profile.d/nix.sh
> 
> in your shell.
> ```

thank you so much

**Comment by bertogg** [comment]
> though note I'm having an issue where I need to start the daemon after each boot. 🤷‍♂️

Hi @OpenBagTwo , this has already been discussed several times in this thread, see here for example: https://github.com/NixOS/nix/issues/7173#issuecomment-1822622214

**Comment by theVakhovskeIsTaken** [comment]
> though note I'm having an issue where I need to start the daemon after each boot. 🤷‍♂️

That's side effects of mounting /etc as an overlay after systemd starts. If Valve mounts /etc overlay in initramfs stage, all of the "user-added" services theoretically could work fine

**Comment by bertogg** [comment]
> If Valve mounts /etc overlay in initramfs stage, all of the "user-added" services theoretically could work fine

This is already happening. The `/etc` overlay is not the problem here and users can add their own systemd services just fine.

It's `/nix` that is only available later during the boot process, and that's where the service file is installed:

```
$ ls -l /etc/systemd/system/nix-daemon.service
lrwxrwxrwx 1 root root 67 Apr 16 14:08 /etc/systemd/system/nix-daemon.service -> /nix/var/nix/profiles/default/lib/systemd/system/nix-daemon.service
```

This is a known limitation: https://github.com/systemd/systemd/issues/8307

You will have the same problem in a regular distro where `/nix` is in a separate filesystem.

Possible solutions are:

- The `daemon-reload` + `start` approach followed by the Determinate Systems installer ([here](https://github.com/DeterminateSystems/nix-installer/blob/v0.12.0/src/planner/steam_deck.rs#L270)).
- That the Nix installer puts an actual file (not a symlink) in `/etc`.

**Comment by bertogg** [comment]
Hi,

with SteamOS 3.6 there are some changes to the OS update system that are relevant for Nix (and other packages that install files in `/etc`).

This should not be a problem for users installing Nix in single-user mode, but multi-user mode is affected.

When an OS update is applied all changes to `/etc` are lost except for some files that are explicitly allowed. The list is available at `/lib/rauc/atomic-update-keep.conf` but users can add additional files by extending `/etc/atomic-update.conf.d/`.

In the case of Nix (using the DeterminateSystems installer) I believe that it would be enough with an `/etc/atomic-update.conf.d/nix.conf` like this:

```
/etc/fish/conf.d/nix.fish
/etc/nix/**
/etc/profile.d/nix.sh
/etc/systemd/system/nix-daemon.socket
/etc/tmpfiles.d/nix-daemon.conf
```

(The installer modifies a couple of other files like `bash.bashrc` but I believe it's better to leave those untouched).

Unfortunately this means that if users are running Nix in SteamOS 3.6 and update their systems the files listed above will be lost and the installation won't work. Fortunately the problem is solvable, and I think we would need a script to repair a broken installation, it would basically need to restore at least `/etc/profile.d/nix.sh`, `/etc/nix/nix-daemon.conf` and the `nix-daemon.socket` symlink.
